# บทที่ 2 พื้นฐานภาษาไพทอน
จากบทที่แล้วที่ได้กล่าวถึงแบบชนิดข้อมูลนามธรรม แต่ก่อนที่จะเริ่มสร้างแบบชนิดข้อมูลนามธรรมอย่างง่าย ผู้พัฒนาจำเป็นต้องมีความรู้ความเข้าใจเกี่ยวกับเครื่องมือหลักก่อน ซึ่งในวิชานี้จะทำการเรียนการสอนด้วยภาษาไพทอนรุ่น 3 ในการสร้างและใช้งานโครงสร้างข้อมูล สาเหตุที่เลือกใช้ภาษาไพทอนเนื่องจากเป็นภาษาที่ทันสมัย เรียนรู้ง่าย เข้าใจง่าย และรองรับการเขียนโปรแกรมเชิงวัตถุ (object-oriented programming) นอกจากนี้ยังมีโครงสร้างข้อมูลแบบอื่นๆ ที่มาพร้อมกับตัวภาษาเองให้เรียกใช้ได้ง่าย 

ให้พิจารณารหัส

In [1]:
print("Hello World")

Hello World


print เป็นฟังก์ชันที่ใช้แสดงผลออกทางหน้าจอ ซึ่งจะเห็นได้ว่าด้วยความสามารถของภาษาไพทอนทำให้การแสดงผลออกหน้าจอทำได้อย่างสะดวกด้วยรหัสเพียงบรรทัดเดียว หากเปรียบเทียบกับภาษาจาวาแล้วผู้พัฒนาจำเป็นต้องเขียนรหัสหลายบรรทัดเพื่อให้ได้ผลลัพธ์แบบเดียวกัน

แต่เนื่องด้วยวิชานี้เป็นวิชาโครงสร้างข้อมูล ไม่ใช่วิชาพื้นฐานการเขียนโปรแกรม ดังนั้นจึงจะขอปูพื้นฐานภาษาไพทอนในส่วนทีจำเป็นเท่านั้น ผู้พัฒนาสามารถเริ่มจากการเข้าไปศึกษาได้ที่ [การสอนภาษาไพทอนฉบับภาษาอังกฤษ](http://docs.python.org/3/tutorial/index.html) [ฉบับภาษาไทย](https://python3.wannaphong.com/%E0%B8%AA%E0%B8%B2%E0%B8%A3%E0%B8%9A%E0%B8%B1%E0%B8%8D/%E0%B8%AA%E0%B8%B3%E0%B8%AB%E0%B8%A3%E0%B8%B1%E0%B8%9A%E0%B8%9C%E0%B8%B9%E0%B9%89%E0%B9%80%E0%B8%A3%E0%B8%B4%E0%B9%88%E0%B8%A1%E0%B8%95%E0%B9%89%E0%B8%99%E0%B8%A0%E0%B8%B2%E0%B8%A9%E0%B8%B2python) หรือ [วิดีทัศน์](https://www.youtube.com/playlist?list=PL0ROnaCzUGB4ieaQndKybT9xyoq2n9NGq)


## 2.1 จุดประสงค์
* เพื่อให้เข้าใจถึงรูปแบบข้อมูลของภาษาไพทอน
* เพื่อให้เข้าใจถึงการรับข้อมูลและการแสดงผลลัพธ์ของภาษาไพทอน

##  2.2 รูปแบบข้อมูล

### 2.2.1 รูปแบบข้อมูลพื้นฐาน
ภาษาไพทอนนั้นจะประกอบด้วยรูปแบบข้อมูลพื้นฐานเชิงตัวเลข 2 ประเภทหลักๆ คือ int และ float กับเชิงตรรกะ (boolean) ซึ่งข้อมูลเหล่านี้จะมีตัวดำเนินการที่ควบคู่มาด้วย แบ่งเป็น 2 ประเภทดังนี้

#### 2.2.1.1 ตัวดำเนินการเชิงคำนวณ (arithmetic operator)
มีความเกี่ยวข้องกับ +, -, \*, / ซึ่งได้กล่าวไว้แล้วใน [บทที่ 1](/notebooks/01-Introduction.ipynb)

#### 2.2.1.2 ตัวดำเนินการเชิงตรรกะ (logical operator)
รูปแบบข้อมูลที่มีความเกี่ยวข้องโดยตรงกับตัวดำเนินการเชิงตรรกะคือ ค่าความจริง (boolean) ประกอบด้วยค่า จริง กับ เท็จ (True กับ False) สำหรับภาษาไพทอนแต่สำหรับภาษาอื่น เช่น ซีจะใช้เลข 0 แทน เท็จ และ 1 หรือเลขอื่นๆ ที่ไม่เท่ากับ 0 แทน จริง การใช้งานตัวดำเนินการเชิงตรรกะนั้นคล้ายคลึงกับภาษาอื่น ขอสรุปด้วย [ตารางที่ 2-1](#table_01)

<a name="table_01"></a>
<center>ตารางที่ 2-1 ตัวดำเนินการเชิงตรรกะ</center>

| ชื่อตัวดำเนินการ      | ตัวดำเนินการ | คำอธิบาย                                                                                                             |
|-------------------|------------|---------------------------------------------------------------------------------------------------------------------|
| น้อยกว่า            |    <       |    x < y เป็นการเปรียบเทียบว่า   x น้อยกว่า y หรือไม่                                                                       |
|    มากกว่า         |    >       |    x > y เป็นการเปรียบเทียบว่า   x มากกว่า y หรือไม่                                                                       |
|    น้อยกว่าเท่ากับ    |    <=      |    x <= y เป็นการเปรียบเทียบว่า   x น้อยกว่าหรือเท่ากับ   y หรือไม่                                                            |
|    มากกว่าเท่ากับ    |    >=      |    x >= y เป็นการเปรียบเทียบว่า   x มากกว่าหรือเท่ากับ   y หรือไม่                                                            |
|    เท่ากับ          |    ==      |    x == y เป็นการเปรียบเทียบว่า   x เท่ากับ y หรือไม่                                                                       |
|    ไม่เท่ากับ        |    !=      |    x != y เป็นการเปรียบเทียบว่า   x ไม่เท่ากับ   y หรือไม่                                                                   |
|    และ            |    and     |    x and y เป็นการเอาค่าความจริง x มา และ   กับ y ซึ่งจะเป็น   True เมื่อ x กับ y เป็น True ทั้งคู่ กรณีอื่นจะเป็น   False หมด      |
|    หรือ            |    or      |    x or y เป็นการเอาค่าความจริง x มา หรือ   กับ y ซึ่งจะเป็น   False เมื่อ x กับ y เป็น False ทั้งคู่   กรณีอื่นจะเป็น True   หมด    |
|    นิเสธ           |    not     |    not x เป็นการกลับค่า   x จาก True เป็น False หรือจาก False เป็น True                                                   |
ตัวอย่างค่าความจริงและตัวดำเนินการเชิงตรรกะมีดังนี้

In [2]:
True

True

In [3]:
False

False

In [4]:
print(True)
print(True and False)
print(not True)
print(not (False or False))

True
False
False
True


ตัวอย่างข้างต้นแสดงให้เห็นว่า คำว่า True คือค่าความจริง True และ คำว่า False ก็คือค่าความจริง False แต่หากไปใช้บนระบบ grader จะต้อง print ออกมาแสดงผลหากโจทย์ต้องการให้แสดง ซึ่งแสดงให้เห็นว่า print(True) จะได้ผลออกมาเป็นค่าความจริง True แต่ไม่ใช่แบบเดียวกับ print("True") และในตัวอย่างอื่นๆ แสดงให้เห็นความสามารถในการใช้ตัวดำเนินการเชิงตรรกะเพื่อให้ได้ผลลัพธ์เป็นค่าความจริง แล้วแสดงผลออกมา เช่น True and False ได้ผลเป็น False สั่ง print จะได้ผลลัพธ์เป็น False ตัวอย่างอื่นๆ นั้นไม่ต่างกัน และสามารถกำหนดค่าความจริงให้กับตัวแปรได้เช่น x กับ y 

In [5]:
x = True
y = False
z = x and y
print(z)

False


ตัวอย่างนี้แสดงเป็นขั้นตอนดังนี้
1. กำหนด x ให้เป็น True
2. กำหนด y ให้เป็น False
3. นำ x มา และ กับ y ได้ค่าความจริง (False) นำมาเก็บไว้ใน z
4. แสดงผลลัพธ์ z ออกทางหน้าจอ

ซึ่งการทำงานจะเหมือนกับ print(True and False) ต่างกันแค่ใช้ตัวแปร (variable) กับไม่ใช้ตัวแปร

#### คำถาม
จากตัวอย่างนี้ให้ทำการแปลง<br />
print(True)<br />
print(not True)<br />
print(not (False or False))<br />
ให้อยู่ในรูปตัวแปรเพื่อจะได้เข้าใจการใช้ตัวแปรอย่างง่ายๆ

In [6]:
# try here

ตัวอย่างที่เป็นการเปรียบเทียบ

In [7]:
print(5==10)
print(10 > 5)
print((5 >= 1) and (5 <= 10))

False
True
True


ตัวอย่างที่ใช้ตัวแปร

In [8]:
x = 5
y = 10
z = (5 == 10)
print(z)

False


บรรทัดที่น่าอาจทำให้สับสนคือ z = (5 == 10) การทำงานเป็นดังนี้คือจะทำการเปรียบเทียบ 5 == 10 ก่อนซึ่งได้ False แล้วกำหนดค่า False ให้กับ z

<br>
<br>
#### คำถาม
จากตัวอย่างนี้ให้ทำการแปลง<br />
print(10 > 5)<br />
print((5 >= 1) and (5 <= 10)))<br />
ให้อยู่ในรูปตัวแปร

In [9]:
# try here

#### คำถาม
จากตัวอย่างการใช้ตัวดำเนินการเชิงตรรกะ ผู้พัฒนาเกิดคำถามอะไรบ้าง

ตัวอย่างการตั้งคำถาม
1. ประกาศชนิดของตัวแปรตรงส่วนใด
2. การเปรียบเทียบนำตัวแปรมาผสมกับตัวเลขได้หรือไม่

คำตอบคือ
1. ภาษาไพทอนไม่จำเป็นต้องประกาศชนิดของตัวแปร แต่ตอนนำไปใช้ต้องระมัดระวังในการใช้ให้ถูกชนิด
2. ทำได้เหมือนภาษาอื่นๆ 

คำถามและคำตอบข้อ 1 จึงได้โยงไปสู่การกำหนดตัวแปรแบบอ้างอิงตัวของภาษาไพทอนดังตัวอย่าง

In [10]:
a = 0
print(a)
a = a + 1
print(a)
a = True
print(a)
a = "Hello"
print(a)

0
1
True
Hello


### 2.2.2 การกำหนดตัวแปรแบบอ้างอิง
ในภาษาไพทอนนั้นการกำหนดตัวแปรจะเป็นแบบอ้างอิงทั้งหมด เปรียบเสมือนการมีชื่อตัวแปรกับลูกศรที่ใช้ชี้ตำแหน่งว่าชื่อตัวแปรนั้นจะชี้ไปยังก้อนข้อมูลตัวใด

<a name="figure_01"></a> 
![alt text](/files/imgs/ref_assign.jpg "assign by reference")
<center>รูปที่ 2-1 การกำหนดตัวแปรแบบอ้างอิง</center>

จากรหัสและ [รูปที่ 2-1](#figure_01)
 เราเริ่มต้นที่ a = 0 สิ่งที่เกิดขึ้นในหน่วยความจำคือ มีก้อนข้อมูล 0 โดยใช้ a เป็นตัวอ้างอิงถึงก้อนข้อมูลนั้น หลังจากนั้นให้ a = a + 1 ซึ่งก้อนข้อมูล 0 จะถูกอ้างอิงแล้วบวกเพิ่มไป 1 แล้วใช้ตัวอ้างอิงตัวเดิมคือ a ทำให้ a ชี้ไปยังก้อนข้อมูลที่เก็บค่า 1 หลังจากนั้นทำการกำหนดค่า a = True ซึ่ง True จะถูกสร้างมาเป็นก้อนข้อมูลใหม่แล้วนำ a มาชี้ที่ก้อนข้อมูลของ True แทน เช่นเดียวกับ a = "Hello"

ด้วยการอ้างถึงข้อมูลแบบอ้างอิงนี้ทำให้ตัวแปร a มีอิสระในการที่จะอ้างอิงข้อมูลชนิดใดก็ได้โดยไม่จำเป็นต้องประกาศเป็นชนิดใดชนิดหนึ่งโดยเฉพาะ

### 2.2.3 รูปแบบโครงสร้างข้อมูล
นอกจากจะมีรูปแบบข้อมูลพื้นฐานแล้วภาษาไพทอนยังมีโครงสร้างข้อมูลแบบง่ายๆ ให้ใช้ได้อีกด้วย เช่น รายการ (list) สายอักขระ (string) เซต (set) หรือ พจนานุกรม (dictionary)

#### 2.2.3.1 รายการ (list)
list เป็นชื่อชนิดรูปแบบข้อมูลหนึ่งในภาษาไพทอน โดยเป็นโครงสร้างข้อมูลที่จะเก็บลำดับของตัวอ้างอิง แต่เพื่อให้ง่ายต่อความเข้าใจให้มองเป็นลำดับของข้อมูลแทน การสร้าง list ทำได้ด้วยการกำหนดค่าของตัวแปรด้วยเครื่องหมาย [ ] และด้วยความพิเศษของการกำหนดตัวแปรแบบอ้างอิงทำให้ข้อมูลใน list ไม่จำเป็นต้องเป็นข้อมูลชนิดเดียวกัน

ตัวอย่าง


In [11]:
myList = []
print(myList)
myList = [1, 3, True, 6.5]
print(myList)

[]
[1, 3, True, 6.5]


บรรทัด myList = [] หมายถึงสร้าง list ว่างๆ ส่วน myList = [1, 3, True, 6.5] หมายถึงการสร้าง list ที่มีข้อมูล 4 ตัวคั่นด้วย , 

ในขณะนี้ให้ผู้พัฒนามองเสมือนตนเองเป็นผู้ใช้โดยที่ภาษาไพทอนเป็นคนจัดหาโครงสร้างข้อมูลให้ เมื่อย้อนกลับไปถึงเรื่องแบบชนิดข้อมูลนามธรรมเราจึงสนใจแค่ว่า list มีตัวดำเนินการใดให้เราใช้ได้บ้าง สรุปได้เป็น [ตารางที่ 2-2](#table_02) ด้านล่างนี้

<br>
<br>
<a name="table_02"></a> 
<center>ตารางที่ 2-2 ตัวดำเนินการของ list</center>

|  ชื่อตัวดำเนินการ    |     ตัวดำเนินการ     |     คำอธิบาย                              |
|-----------------|------------------|----------------------------------------------|
| ดัชนี             | [ ]              | เข้าถึงข้อมูลด้วยดัชนีที่อยู่ใน [ ]                     |
| การต่อ           | +                | การต่อ list 2 ตัวเข้าด้วยกัน                      |
| การซ้ำ           | *                | การต่อ list ด้วยการซ้ำสมาชิกของตัวเอง             |
| การเป็นสมาชิก     | in               | x in list การถามว่า x เป็นสมาชิกใน list หรือไม่   |
| ความยาว         | len              | จำนวนข้อมูลใน list                             |
| การแบ่งส่วน       | [ : ]            | [i:j] การดึงมาเฉพาะข้อมูลตัวที่ i ถึง j-1           |

ตัวอย่าง

In [12]:
myList = [0,1,2,3,4,5]
print(myList[0])
myList[0] = 10
print(myList)
A = myList*3
print(A)
print(len(A))
print(A[2:5])
print(A[1:5:1])
print(A[1:5:2])
print(4 in A)
B = [4] * 3
print(B)
print(4 in B)

0
[10, 1, 2, 3, 4, 5]
[10, 1, 2, 3, 4, 5, 10, 1, 2, 3, 4, 5, 10, 1, 2, 3, 4, 5]
18
[2, 3, 4]
[1, 2, 3, 4]
[1, 3]
True
[4, 4, 4]
True


การประกาศและการเข้าถึงด้วยดัชนี (index) เป็นสิ่งที่ผู้พัฒนาคุ้นเคยอยู่แล้ว ซึ่งภาษาไพทอนจะเหมือนภาษาซีกับจาวา ที่ดัชนีเริ่มที่ 0 แต่ที่แตกต่างออกไปคือ A = myList\*3 ซึ่งจะทำการซ้ำ myList ไป 3 รอบ ได้ผลลัพธ์อะไรส่งต่อให้กับ A สิ่งที่ได้คือ list ที่มีสมาชิกของ myList ต่อกันไป 3 ชุด อีกหนึ่งตัวดำเนินการที่น่าสนใจคือ 4 in A จะทำการตรวจสอบว่ามี 4 อยู่ใน A หรือไม่ ซึ่งคำตอบคือ True เพราะมีอยู่

#### คำถาม
ตัวดำเนินการ [:] และ [::] มีการใช้งานอย่างไร (ให้ทดลองค้นหาด้วยคำสำคัญ python list slicing)

#### คำถาม
ให้ผู้พัฒนาพิจารณาว่า ถ้า [1,2,3,4] \* 2 ได้ผลลัพธ์เป็น [1,2,3,4,1,2,3,4] แล้ว [[1,2,3,4]] \* 2 จะได้ผลลัพธ์เป็นอะไร


In [13]:
# try here

คำตอบคือ

เมื่อคิดเป็นขั้นเป็นตอนแบบใช้หลักการและเหตุผล ถ้า [1,2,3,4] \* 2 ได้ output เป็น [1,2,3,4,1,2,3,4] เหมือนทำการแกะเปลือก [ ] ออก ไส้ในจะเหลือ 1,2,3,4 แล้วทำสำเนามาอีก 1 ชุดเพื่อมาต่อท้ายได้เป็น 1,2,3,4,1,2,3,4 แล้วจึงใส่เปลือก [ ] กลับเข้าไป <br />
ดังนั้นถ้าเป็น [[1,2,3,4]] \* 2 ให้ทำแบบเดียวกันคือแกะเปลือก [ ] ออก ไส้ในจะเหลือ [1,2,3,4] ทำสำเนามาอีก 1 ชุดเพื่อต่อท้ายได้เป็น [1,2,3,4],[1,2,3,4] แล้วใส่เปลือก [ ] กลับเข้าไปได้เป็น [[1,2,3,4],[1,2,3,4]] ซึ่งคือ list ซ้อน list อีกชั้นหนึ่ง หรือจะมองกันแบบง่ายๆ ก็คือ อาร์เรย์ (array) 2 มิติเป็น

แถวที่ 0 [1,2,3,4]<br>
แถวที่ 1 [1,2,3,4]

#### คำถาม
สมมติว่ามี list ซ้อน list, myList = [[1,2,3,4],[5,6,7,8],[9,10,11,12]] ถ้าต้องการเข้าถึงด้วยดัชนีที่ค่า 10 จะเข้าถึงได้อย่างไร ให้ทดลองด้วยตัวเองหรือค้นคว้าด้วยคำสำคัญ "python 2d array"


In [14]:
# try here

คำตอบคือ


ลองวาดให้อยู่ในรูปอย่างง่ายก่อน <br>
แถวที่ 0 [1,2,3,4]<br>
แถวที่ 1 [5,6,7,8]<br>
แถวที่ 2 [9,10,11,12]<br>
myList\[2\]\[1\] เพราะอยู่แถวที่ 2 และอยู่คอลัมน์ 1 ดังนั้นถ้าเจอ list ซ้อน list ให้พยายามจัดรูปอย่างง่ายก่อนแล้วจึงพิจารณาดัชนีเป็นแถวและคอลัมน์ตามลำดับ และคำนึงเสมอว่าดัชนีจะต้องเริ่มที่ 0 

มีอีกหนึ่งประเด็นที่จะต้องระวังสืบเนื่องจากการกำหนดข้อมูลแบบอ้างอิง ลองพิจารณาจากตัวอย่างด้านล่าง

In [15]:
myList = [1,2,3,4]
A = myList
print(A)
A[0] = 10
print(myList)

[1, 2, 3, 4]
[10, 2, 3, 4]


เห็นได้ว่าหลังจากที่เรากำหนด A = myList แล้ว เมื่อมีการเปลี่ยนแปลงค่าใน A ค่าใน myList จะเปลี่ยนแปลงตามไปด้วย สาเหตุที่เป็นเช่นนี้ก็เพราะว่า A กับ myList ถูกกำหนดให้ชี้ไปยังก้อนข้อมูลก้อนเดียวกัน ดังนั้นไม่ว่าจะเปลี่ยนแปลงค่าผ่าน A หรือ myList ก็จะเสมือนว่าข้อมูลของอีกตัวแปรที่เหลือจะถูกเปลี่ยนตามไปด้วย ดังแสดงตัวอย่างใน [รูปที่ 2-2](#figure_02)

<a name="figure_02"></a>

![alt text](/files/imgs/ref_assign_list.jpg "list assignment")
<center>รูปที่ 2-2 การกำหนดค่าใน list</center>

อย่างไรก็ดีข้อระวังนี้จะเกิดขึ้นกับรูปแบบโครงสร้างข้อมูลเท่านั้น ส่วนรูปแบบข้อมูลพื้นฐานจะไม่เป็นปัญหาดังตัวอย่าง

In [16]:
a = 5
b = a
b = 4
print(a)
print(b)

5
4


นอกจาก list จะมีตัวดำเนินการให้แล้ว อีกสิ่งหนึ่งที่เป็นตัวดำเนินเหมือนกันแต่อยู่ในรูปของฟังก์ชั่นก็คือ เมธอด (method) ซึ่งคล้ายกับภาษาจาวา โดยเมธอดของ list ที่มีให้ใช้ดัง [ตารางที่ 2-3](#table_03)

<a name="table_03"></a>
<center>ตารางที่ 2-3 เมธอดของ list</center>

| ชื่อเมธอด | การใช้งาน             | คำอธิบาย                                |
|-----------|----------------------|----------------------------------------|
| append    | alist.append(item)   | เพิ่ม item ต่อข้างท้ายของ alist             |
| insert    | alist.insert(i,item) | ใส่ item เข้าไปที่ตำแหน่งที่ i ของ alist      |
| pop       | alist.pop()          | ดึงข้อมูลตัวสุดท้ายของ alist ออกมาพร้อมทั้งคืนค่า |
| pop       | alist.pop(i)         | ดึงข้อมูลตัวที่ i ของ alist ออกมาพร้อมทั้งคืนค่า  |
| sort      | alist.sort()         | แก้ไข alist ให้เรียงลำดับ                  |
| reverse   | alist.reverse()      | แก้ไข alist ให้กลับลำดับ                   |
| del       | del alist[i]         | ลบข้อมูลตัวที่ i                            |
| index     | alist.index(item)    | คืนค่าดัชนีแรกที่ค้นเจอ item                  |
| count     | alist.count(item)    | นับจำนวน item ที่เจอใน alist              |
| remove    | alist.remove(item)   | ลบข้อมูล item ที่เจอตัวแรกใน alist          |

การเรียกใช้เมธอดก็ใช้แบบเดียวกับภาษาจาวาคือใช้ . ตัวอย่างการใช้งานเมธอดของ list มีดังนี้

In [17]:
myList = [1024, 3, True, 6.5]
myList.append(False)
print(myList)
myList.insert(2,4.5)
print(myList)
print(myList.pop())
print(myList)
print(myList.pop(1))
print(myList)
myList.pop(2)
print(myList)
myList.sort()
print(myList)
myList.reverse()
print(myList)
print(myList.count(6.5))
print(myList.index(4.5))
myList.remove(6.5)
print(myList)
del myList[0]
print(myList)

[1024, 3, True, 6.5, False]
[1024, 3, 4.5, True, 6.5, False]
False
[1024, 3, 4.5, True, 6.5]
3
[1024, 4.5, True, 6.5]
[1024, 4.5, 6.5]
[4.5, 6.5, 1024]
[1024, 6.5, 4.5]
1
2
[1024, 4.5]
[4.5]


#### คำถาม
มีเมธอดใดบ้างที่ทำการแก้ไข list จงแสดงตัวอย่าง

In [18]:
# try here

นอกจากเมธอดแล้วฟังก์ชันพื้นฐานของภาษาไพทอนที่มักจะใช้ร่วมกับ list คือ range เป็นฟังก์ชันที่เอาไว้สร้างชุดของตัวเลข เช่น

In [19]:
print(range(10))
print(list(range(10)))
print(list(range(5,10)))
print(list(range(5,10,1)))
print(list(range(5,10,2)))
print(list(range(10,5)))
print(list(range(10,5,-1)))
print(list(range(10,5,-2)))

range(0, 10)
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[5, 6, 7, 8, 9]
[5, 6, 7, 8, 9]
[5, 7, 9]
[]
[10, 9, 8, 7, 6]
[10, 8, 6]


จากตัวอย่างจะเห็นได้ว่า range ให้ผลลัพธ์ออกมาเป็น object ชนิด range จึงต้องแปลงให้อยู่ในรูปของ list ก่อน สิ่งที่ปรากฎอยู่บนเอกสารคู่มือภาษาไพทอนคือ

range(start, stop[, step])

ให้ผลออกมาเป็นลำดับของตัวเลขตั้งแต่ start ถึง stop-1 ด้วยการกระโดดข้ามเป็นจำนวน step คำอธิบายนี้อาจจะยังไม่ชัดเจนให้ดูตัวอย่างด้านบนจะเห็นว่า step จะมีหรือไม่ก็ได้ แต่ถ้าไม่มีจะหมายความว่า step เป็น 1 นอกจากนี้ถ้า start เป็นค่าที่มากกว่า stop ถ้าไม่กำหนด step จะไม่มีอะไรเกิดขึ้นใน list เลย ดังนั้นจึงจำเป็นต้องกำหนด step เป็นค่าที่ติดลบเพื่อให้กระโดดถอยหลัง

#### คำถาม
1. จงสร้าง list ที่มีค่าตั้งแต่ 50 - 60 ด้วย range 
2. จงสร้าง list ที่มีค่าตั้งแต่ 60 - 50 ด้วย range
3. ข้อระวังในการใช้ range คืออะไร

In [20]:
# try here

#### 2.2.3.2 สายอักขระ (string)
สายอักขระในภาษาไพทอนจะแทนด้วย str เปรียบเป็นโครงสร้างข้อมูลอย่างหนึ่ง โดยเก็บเป็นลำดับของตัวอักษร (character) ซึ่งประกอบด้วยตัวอักษร ตัวเลข หรือสัญลักษณ์อื่นๆ โดย str จะประกาศด้วยสัญลักษณ์ " หรือ ' ต่างจากภาษาจาวาที่ใช้ ' สำหรับตัวอักษร

In [21]:
s0 = "i am the first string."
s1 = 'i am the second string.'
print(s0)
print(s1)
print(s0[2])
print(s1 * 2)
print(len(s0))

i am the first string.
i am the second string.
a
i am the second string.i am the second string.
22


เห็นได้ว่า str มีลักษณะข้อมูลแบบเป็นลำดับ ดังนั้นจึงใช้งานหลักๆ ได้เหมือนกับ list แต่ก็มีเมธอดเฉพาะตัวเช่นกันดัง [ตารางที่ 2-4](#table_04)

<a name="table_04"></a>
<center>ตารางที่ 2-4 เมธอดของ str</center>

| ชื่อเมธอด | การใช้งาน             | คำอธิบาย                                              |
|-----------|----------------------|------------------------------------------------------|
| center    | astring.center(w)    | คืนค่า str ที่วางตัวอยู่ตรงกลางเมื่อกำหนดช่วงความยาว w      |
| count     | astring.count(item)  | คืนค่าการนับจำนวน item ที่ปรากฎใน astring                 |
| ljust     | astring.ljust(w)     | คืนค่า str ที่วางตัวชิดซ้ายเมื่อกำหนดช่วงความยาว w          |
| lower     | astring.lower()      | คืนค่า str ที่แปลงเป็นตัวอักษรตัวเล็ก                      |
| upper     | astring.upper()      | คืนค่า str ที่แปลงเป็นอักษรตัวใหญ่                        |
| rjust     | astring.rjust(w)     | คืนค่า str ที่วางตัวชิดขวาเมื่อกำหนดช่วงความยาว w          |
| find      | astring.find(item)   | คืนค่าดัชนีแรกที่เจอ item                                  |
| split     | astring.split(schar) | แบ่ง astring ด้วยตัวคั่น schar แล้วคืนค่าเป็น list ของ str |

In [22]:
s0_upper = s0.upper()
print(s0_upper)
print(s0.count("t"))
s2 = "1 2 3 4 5 6"
print(s2.split(' '))
print(s2.split())
s3 = "a,b,c,d,e,f"
print(s3.split(','))

I AM THE FIRST STRING.
3
['1', '2', '3', '4', '5', '6']
['1', '2', '3', '4', '5', '6']
['a', 'b', 'c', 'd', 'e', 'f']


จากตัวอย่างแสดงให้เห็นถึงการใช้งานอย่างง่าย เมธอดที่จะใช้งานบ่อย คือ split เห็นได้ว่า split(' ') กับ split() มีพฤติกรรมแบบเดียวกันคือใช้ space เป็นตัวคั่น แต่มีข้อสังเกตคือ

<br>
<br>
#### คำถาม
1. หากต้องการแบ่งตัวอักษรทุกตัวใน str ให้อยู่ในรูป list ของตัวอักษรต้องทำอย่างไร ให้ค้นหาแล้วบอกว่าคำสำคัญที่ใช้คืออะไร แล้วได้คำตอบว่าอะไร
2. ถ้ากำหนด s = "ttttttttt" จะนับ "ttt" ด้วย s.count("ttt") ได้ผลลัพธ์คืออะไร ให้คิดก่อนแล้วจึงหาคำตอบด้วยการเขียนรหัส

In [23]:
# try here

In [24]:
s = "hello"
print(s[0])
s[0] = "y"

h


TypeError: 'str' object does not support item assignment

จากตัวอย่างพบว่าพฤติกรรมที่แตกต่างจาก list คือ str สามารถดึงค่าด้วยดัชนีได้ แต่ไม่สามารถแก้ไขค่าด้วยดัชนีได้ ดังนั้นถ้าหากต้องการแก้ไข str จะต้องแปลง str ให้เป็น list ก่อน แล้วแก้ไขตามดัชนีที่ต้องการแล้วจึงแปลงกลับไปเป็น str ดังนี้

In [25]:
s = 'hello'
print(s)
l = list(s)
l[0] = "y"
s = ''.join(l)
print(s)

hello
yello


#### คำถาม
จากตัวอย่างด้านบน คำสั่งใดที่ทำการแปลงจาก str ให้อยู่ในรูป list ของตัวอักษร และคำสั่งใดที่ทำการแปลง list ของตัวอักษร กลับไปเป็น string

In [26]:
# try here

<br>
#### 2.2.3.3 เซต (set)
set เป็นโครงสร้างข้อมูลที่ไม่มีลำดับ หมายความว่าเมื่อใส่ข้อมูลเข้าไปใน set เป็นลำดับแล้วเมื่อเรียกใช้เป็นลำดับอาจได้ผลออกมาไม่เหมือนกับลำดับตอนที่ใส่ข้อมูล

In [27]:
mySet = {3,6,"cat",4.5,False}
print(mySet)

{False, 3, 4.5, 6, 'cat'}


ดังตัวอย่างก็ชัดเจนว่าใส่ข้อมูลแบบหนึ่งแต่ผลลัพธ์ที่ได้ออกมาเป็นอีกแบบ และการสร้าง set เราจะใช้เครื่องหมาย { } และมี , เป็นตัวคั่นแต่ละสมาชิก เนื่องจาก set เป็นโครงสร้างข้อมูลที่แทบจะไม่ได้ถูกใช้เลยในรายวิชานี้ จึงขอทิ้งไว้เป็นการบ้าน

#### คำถาม
จงสืบค้นเพื่ออธิบายว่า set มีตัวดำเนินการและเมธอดอะไรให้ใช้บ้าง และแสดงตัวอย่างการใช้งานแต่ละตัวดำเนินการและเมธอด

In [28]:
# try here

#### 2.2.3.4 พจนานุกรม (dictionary)
พจนานุกรมในภาษาไพทอนจะแทนด้วย dict เป็นโครงสร้างข้อมูลของคู่ลำดับ ตัวหนึ่งคือ คีย์ (key) และอีกตัวคือ ค่า (value) การสร้าง dict นั้นอยู่ภายในสัญลักษณ์ { } เราแทนคู่ลำดับ key : value คั่นด้วย : คือ 1 ข้อมูลใน dict แล้วคั่นแต่ละข้อมูลด้วย , 

In [29]:
capitals = {'Iowa':'DesMoines','Wisconsin':'Madison'}
print(capitals)

{'Iowa': 'DesMoines', 'Wisconsin': 'Madison'}


เราสามารถเข้าถึง value ใน dict ได้ด้วย key

In [30]:
print(capitals['Iowa'])

DesMoines


การเข้าถึงข้อมูลจะคล้ายๆ กับ list ที่ใช้ดัชนีเป็นตัวเลข แต่ dict จะใช้ key แทนดัชนี นอกจากการเข้าถึงข้อมูลแล้ว การเพิ่มข้อมูลก็ยังใช้รูปแบบเดียวกันคือ

In [31]:
capitals['Utah']='SaltLakeCity'
capitals['California']='Sacramento'
capitals['Thailand'] = 'Bangkok'
print(capitals)

{'Iowa': 'DesMoines', 'Wisconsin': 'Madison', 'Utah': 'SaltLakeCity', 'California': 'Sacramento', 'Thailand': 'Bangkok'}


dict มีตัวดำเนินการและเมธอด ดัง [ตารางที่ 2-5](#table_05) และ [ตารางที่ 2-6](#table_06)

<br>
<a name="table_05"></a>
<center>ตารางที่ 2-5 ตัวดำเนินการของ dict</center>

| ตัวดำเนินการ | การใช้งาน       | คำอธิบาย                                         |
|----------|----------------|-------------------------------------------------|
| []       | myDict[k]      | คืนค่า value ที่เป็นคู่ของ key k                       |
| in       | key in adict   | คืนค่า True เมื่อมี key อยู่ใน adict ถ้าไม่มีจะคืนค่า False |
| del      | del adict[key] | ลบคู่ลำดับของ key ออกจาก adict                     |


<a name="table_06"></a>
<center>ตารางที่ 2-6 เมธอดของ dict</center>

| ชื่อเมธอด | การใช้งาน         | คำอธิบาย                                          |
|-----------|------------------|--------------------------------------------------|
| keys      | adict.keys()     | คืนค่า key ทั้งหมดในรูปแบบข้อมูล dict_keys              |
| values    | adict.values()   | คืนค่า value ทั้งหมดในรูปแบบข้อมูล dict_values          |
| items     | adict.items()    | คืนค่าคู่ลำดับ key-value ทั้งหมดในรูปแบบข้อมูบ dict_items  |
| get       | adict.get(k)     | คืนค่า value ที่สัมพันธ์กับ key k ถ้าไม่เจอจะคืนค่า None แทน |
| get       | adict.get(k,alt) | คืนค่า value ที่สัมพันธ์กับ key k ถ้าไม่เจอจะคืนค่า alt แทน  |

In [32]:
phoneext={'david':1410,'brad':1137}
print(phoneext)
print(phoneext.keys())
print(list(phoneext.keys()))
print(phoneext.values())
print(list(phoneext.values()))
print(phoneext.items())
print(list(phoneext.items()))
print(phoneext.get("kent"))
print(phoneext.get("kent","NO ENTRY"))

{'david': 1410, 'brad': 1137}
dict_keys(['david', 'brad'])
['david', 'brad']
dict_values([1410, 1137])
[1410, 1137]
dict_items([('david', 1410), ('brad', 1137)])
[('david', 1410), ('brad', 1137)]
None
NO ENTRY


จากตัวอย่างจุดสังเกตหลักคือการคืนค่ามาเป็นรูปแบบข้อมูล dict_keys, dict_values และ dict_items หากต้องการนำมาใช้งานต่อจะต้องแปลงให้อยู่ในรูปของ list ก่อน และ get สามารถใช้ได้ 2 แบบคือแบบที่หา key ไม่เจอแล้วคืนค่า None หรือคืนค่าที่ผู้ใช้สามารถกำหนดได้เอง

## 2.3 ข้อมูลขาเข้าและการแสดงผลลัพธ์
ถือเป็นสิ่งที่สำคัญอย่างหนึ่งเพราะการใช้งานระบบ grader ต้องรับข้อมูลขาเข้าให้ถูกรูปแบบ รวมถึงแสดงผลลัพธ์ให้ถูกรูปแบบด้วยเช่นกัน

การรับข้อมูลจะรับผ่านทางแป้นพิมพ์โดยฟังก์ชันที่เกี่ยวข้องก็ตรงตัวเลยคือ input มีการเรียกใช้ดังนี้

In [33]:
aName = input('Please enter your name: ')
print(aName)
aName = input()
print(aName)

Please enter your name: John
John
Jane
Jane


---
**ข้อสังเกต**

จากตัวอย่างจะเห็นได้ว่าถ้าหากส่ง str เข้าไปเป็นพารามิเตอร์ให้กับ input() จะทำพิมพ์ออกมาทางหน้าจอด้วย str นั้นก่อนแล้วจึงรับข้อมูล ซึ่งต้องระวังเพราะในการใช้งานระบบ grader นั้น ไม่ต้องการให้แสดงผลอะไรที่ไม่จำเป็นออกมา จึงมักจะใช้ input() แบบไม่รับพารามิเตอร์เป็นหลัก โดยการทำงานของโปรแกรมจะรอรับข้อมูลโดยที่ไม่แสดงผลอะไรออกมาก่อน

---

สิ่งที่ต้องคำนึงถึงอีกเรื่องหนึ่งคือ ค่าที่ได้มาจาก input() คือ str แม้จะรับมาเสมือนเป็นตัวเลขแต่แท้จริงแล้วคือ str ดังนั้นถ้าต้องการนำมาคำนวณจะต้องแปลงให้เป็นชนิดข้อมูลแบบตัวเลขก่อน เช่น

In [34]:
a = input()
b = input()
print(a+b)
c = int(a)
d = int(b)
print(c+d)
e = float(a)
f = float(b)
print(e+f)

1
2
12
3
3.0


ถ้ารับ a กับ b แล้วนำมา + กันโดยตรง ผลที่ได้คือการ + str ซึ่งคือการต่อ (concatenation) str แต่ถ้าแปลงให้เป็น int ก่อนด้วยฟังก์ชัน int() หรือ แปลงให้เป็น float ด้วยฟังก์ชัน float() ผลของการ + จะได้เป็นไปตามรูปแบบข้อมูลนั้นๆ 

### 2.3.1 รูปแบบของสายอักขระ (string format)
ที่ผ่านมาผู้พัฒนาได้เรียนรู้การใช้ print เพื่อแสดงผลตัวแปรตรงๆ โดยรูปแบบจะเป็นไปตามภาษาไพทอนแต่เราก็สามารถที่จะแสดงผลในลักษณะอื่นได้อีก เช่น

In [35]:
print("Hello")
print("Hello","World")
print("Hello","World", sep="***")
print("Hello","World", end="xxx")
print("end")

Hello
Hello World
Hello***World
Hello Worldxxxend


print ถ้าคั่นด้วย , จะแสดงผลออกมามี space คั่นให้โดยอัตโนมัติ แต่ถ้าไม่ต้องการได้ space สามารถแทนด้วย str อื่นได้ ในตัวอย่างนี้คือ \*\*\* โดยกำหนดผ่านพารามิเตอร์ชื่อ sep และสิ่งที่ print ทำให้ตลอดเวลาคือแสดงผลแล้วจะขึ้นบรรทัดใหม่ ซึ่งให้ทำความเข้าใจว่าคือ str ที่แทนการขึ้นบรรทัดใหม่ แต่ถ้าต้องการให้จบด้วย str แบบอื่นสามารถทำได้เช่นกันผ่านการกำหนดผ่านพารามิเตอร์ชื่อ end ดังตัวอย่างนี้คือ xxx จะเห็นว่าไม่ขึ้นบรรทัดใหม่แต่จะลงท้ายด้วย xxx แทน

#### คำถาม
หากต้องการแสดงผล str 2 ตัวติดกันโดยไม่ใช้ตัวดำเนินการ + แต่ใช้ผ่านพารามิเตอร์ sep จะต้องทำอย่างไร

In [36]:
# try here

นอกจากนี้ยังสามารถกำหนดรูปแบบของ str ได้เอง โดยใช้เครื่องหมาย %

In [37]:
aName = "John"
age = 20
print("%s is %d years old." % (aName, age))

John is 20 years old.


% ที่อยู่ภายใน " " บ่งบอกถึงรูปแบบข้อมูลที่จะนำมาแสดงผล ในขณะที่ % ที่อยู่ข้างนอกบ่งบอกถึงลำดับของตัวแปรที่จะนำมาแสดงผล ตามตัวอย่างนี้คือ %s เจอเป็นลำดับแรกจะนำ aName มาแสดงผล %d เจอเป็นลำดับที่สองจะนำ age มาแสดงผล

ข้อสังเกตคือ %s %d คืออะไร แท้จริงแล้วภาษาจาวาก็มีรูปแบบของสายอักขระ เช่นกัน [รูปแบบของสายอักขระสำหรับภาษาจาวา](https://dzone.com/articles/java-string-format-examples) ภาษาไพทอนก็เช่นเดียวกัน %s แทนการแสดงผลแบบ str และ %d แทนการแสดงผลแบบ int มีการแสดงผลแบบอื่นๆ อีกดังแสดงใน [ตารางที่ 2-7](#table_07)

<a name="table_07"></a>
<center>ตารางที่ 2-7 รูปแบบของ str</center>

| ตัวอักษร | รูปแบบข้อมูล                              |
|-----------|----------------------------------------|
| d, i      | int                                    |
| u         | int แบบไม่มีเครื่องหมาย (unsigned integer) |
| f         | float ในรูปแบบ m.ddddd                  |
| e         | float ในรูปแบบ m.ddddde+/-xx            |
| E         | float ในรูปแบบ m.dddddE+/-xx            |
| c         | ตัวอักษร 1 ตัว                         |
| s         | str                                 |
| %         | % โดยเวลาใช้คือ %% จะแสดงผลเป็น %         |

การใช้งานเป็นหลักคือ d f s ซึ่งตัวที่จะก่อปัญหาคือ f เพราะ

In [38]:
a = 2.6666
print("%f" % (a))
print("%.2f" % (a))

2.666600
2.67


จากตัวอย่างจะเห็นว่ามีการเพิ่ม .2 เข้าไประหว่าง % และ f ผลที่ได้คือแสดงผลออกมาเป็น float ที่มีทศนิยม 2 ตำแหน่งและปัดเศษ ทั้งนี้สิ่งที่ต้องคำนึงถึงคือการคำนวณ float จะมีสิ่งที่เรียกว่า ความผิดพลาดจากการปัดเศษ (round off error) จะเห็นได้ว่า 2.666600 มีความถูกต้องมากกว่า 2.67 ซึ่งความผิดพลาดนี้เกิดขึ้นได้เสมอเมื่อมีการคำนวณ float หลายๆ ครั้ง ดังนั้นในการใช้ระบบ grader โจทย์จะกำหนดว่าผลลัพธ์สุดท้ายต้องการทศนิยม 2 หรือ 3 ตำแหน่งเพื่อหลีกเลี่ยงปัญหาจากความผิดพลาดนี้ ดังนั้นแสดงผล float ที่มีทศนิยม 2 ตำแหน่งใช้ %.2f และ 3 ตำแหน่งใช้ %.3f

In [39]:
a = 2.6666
print("%.2f" % (a))
print("%.3f" % (a))

2.67
2.667


หลักๆ ที่จะใช้ในรายวิชาจะมีเพียงเท่านี้ 

## 2.4 โครงสร้างควบคุม (control structure)
ในการเขียนโปรแกรมโดยเฉพาะภาษาจาวานั้นจะต้องมีการตัดสินใจและการใช้วงวนอยู่ตลอด สิ่งต่างๆ เหล่านี้เรียกว่า โครงสร้างควบคุม ซึ่งภาษาไพทอนมี while และ for ให้ใช้สำหรับวงวน แต่ วากยสัมพันธ์ (syntax) จะต่างกัน

### 2.4.1 while

In [40]:
counter = 0
while counter < 5 :
    print("hello", counter)
    counter = counter + 1

hello 0
hello 1
hello 2
hello 3
hello 4


จากตัวอย่างจะทำวงวนของการแสดงผล hello กับตัวนับ counter จำนวน 5 รอบ จาก 0 - 4 ข้อสังเกตคือถ้าต้องการทำ 5 รอบ เหตุใดจึงไม่นับจาก 1 - 5 เจตนาเพราะการใช้ดัชนีของภาษาไพทอนที่เริ่มที่ 0 ถ้าต้องการทำวงวน n รอบ ให้วนจนตัวนับไม่เกิน n หรือสิ้นสุดที่ n - 1 ในกรณีนี้คือ วนจนตัวนับไม่เกิน 5 หรือสิ้นสุดที่ 5 - 1 = 4

การไล่เรียงรหัสอย่างละเอียดเป็นดังนี้
1. เริ่มต้นค่า counter = 0 
2. ก่อนที่จะทำวงวน while จะทำการตรวจสอบเงื่อนไขว่า counter < 5 หรือไม่ พบว่า 0 < 5 เป็น True เข้าไปทำในวงวน
3. แสดงผล hello ตามด้วย counter ได้เป็น hello 0
4. เพิ่มค่า counter ไป 1 ณ ขณะนี้ counter = 1
5. ก่อนที่จะทำวงวน while จะทำการตรวจสอบเงื่อนไขว่า counter < 5 หรือไม่ พบว่า 1 < 5 เป็น True เข้าไปทำในวงวน
6. แสดงผล hello ตามด้วย counter ได้เป็น hello 1
7. เพิ่มค่า counter ไป 1 ณ ขณะนี้ counter = 2
ซ้ำไปเรื่อยๆ จน counter = 5
8. ก่อนที่จะทำวงวน while จะทำการตรวจสอบเงื่อนไขว่า counter < 5 หรือไม่ พบว่า 5 < 5 เป็น False แล้วจึงไม่เข้าวงวน และจบการทำงาน

เมื่อพิจารณาถึงเรื่องของวากยสัมพันธ์พบว่า สำหรับ while จะตามด้วยข้อความสั่ง (statement) ที่บอกถึงเงื่อนไข (condition) ซึ่งเงื่อนไขจะถูกทดสอบ หากได้ผลเป็น True จะเข้าไปทำวงวน แต่จะสังเกตได้ว่าถ้าเป็นภาษาจาวาจะมี { } บอกขอบข่าย (scope) ของวงวน แต่ในภาษาไพทอนจะใช้ย่อหน้า (indent) เป็นตัวบอกขอบข่ายแทน นั่นคือ 

 print("hello", counter) <br>
 counter = counter + 1 
 
อยู่ในขอบข่ายของวงวน while ประเด็นยิบย่อยแต่มักทำให้เกิดข้อผิดพลาดคือ เครื่องหมาย : ที่เอาไว้ปิดเงื่อนไข ซึ่งผู้พัฒนาควรต้องระวังหรือแปลความหมายความผิดพลาดจากวากยสัมพันธ์ (syntax error) ที่ภาษาไพทอนฟ้องมา แล้วแก้ไขให้ถูกต้อง ส่วนจุดดีของภาษาไพทอนคือไม่จำเป็นต้องใช้ ; จบคำสั่ง ให้พิจารณาตัวอย่างเพิ่มเติม

In [41]:
alist = [10,9,8,7,6,5,4]
wannafind = 7
counter = 0
found = False
length = len(alist)
while counter < length and not found :
    found = (wannafind == alist[counter])
    print(counter, found)
    counter = counter + 1

0 False
1 False
2 False
3 True


ตัวอย่างนี้มีสิ่งที่เพิ่มเข้ามาคือตัวแปร found เริ่มต้นเป็น False ซึ่งจะมีการเปลี่ยนแปลงเป็น True เมื่อ wannafind มีค่าเท่ากับ alist ดัชนีที่ counter ตามรหัส

found = (wannafind == alist[counter])

ซึ่ง found ถูกนำไปใช้เพิ่มเงื่อนไขสำหรับ while คือถ้าเป็นค่าเริ่มต้น False เราจะได้ว่า not found หรือ not False คือ True นำมา and กับเงื่อนไข counter < length ในเชิงความหมายคือ จะทำวงวนเมื่อ counter ไม่เกินความยาว length และ ยังไม่ found เงื่อนไขจะเป็น False เมื่อ counter สุดที่ length หรือ found แล้วเป็น True

#### คำถาม
หาก wannafind = 1 วงวน while จะสิ้นสุดที่ counter มีค่าเท่าใด และ found มีค่าอะไร

### 2.4.2 for
สำหรับ for แล้วจะค่อนข้างต่างจากภาษาจาวา ซึ่งโดยปกติมักจะใช้ for กับการไล่ลำดับของตัวเลข แต่สำหรับภาษาไพทอนแล้ว for จะทำการหยิบข้อมูลจาก list, range หรือ str มาทีละค่า

In [42]:
alist = [0,1,2,3,4,5]
for item in alist:
    print(item * 2)

0
2
4
6
8
10


In [43]:
r = range(6)
for item in r:
    print(item * 2)


0
2
4
6
8
10


In [44]:
for item in range(6):
    print(item *2)

0
2
4
6
8
10


In [45]:
s = "012345"
for item in s:
    print(item * 2)

00
11
22
33
44
55


In [46]:
s = "012345"
for item in s:
    print(int(item) * 2)

0
2
4
6
8
10


ตัวอย่างทั้ง 5 for จะหยิบค่าจาก list, range หรือ str โดยให้ค่า item ที่เกือบจะเหมือนกันทั้งหมดคือ int จาก 0 - 5 แต่ในกรณีที่เป็น string จะแตกต่างตรงที่ for จะหยิบตัวอักษรทีละตัว ถ้าจะนำค่ามาใช้ในรูปแบบ int จะต้องแปลงด้วย int() ก่อน

เมื่อทราบแล้วว่า for สามารถใช้หยิบค่าทีละตัวจาก list ได้ และหยิบทีละตัวอักษรจาก str ได้ หากเรามี list ของ str แล้วต้องการแยกตัวอักษรออกมาอีกเป็นตัวๆ จะแก้ได้ด้วย for ซ้อน for ดังตัวอย่าง

In [47]:
ls = ["ant","bee","cat","dog"]
for astring in ls :
    print("The string is",astring)
    for achar in astring :
        print(achar)

The string is ant
a
n
t
The string is bee
b
e
e
The string is cat
c
a
t
The string is dog
d
o
g


โดย for แรกสุดและอยู่วงนอกสุดจะทำการไล่หยิบทีละ str จาก list เก็บไว้ใน astring แล้วแสดงผลออกมาให้ดูก่อน แล้ว for ที่สองที่อยู่วงในทำการหยิบแต่ละ ตัวอักษรใน astring ใส่ใน achar แล้วจึงแสดงผล

### 2.4.3 if else elif
จะเหมือนกับของภาษาจาวา คือ เป็นทางเลือกในการตัดสินใจ

In [48]:
import math

n = int(input())
if n < 0:
   print("Negative value cannot find the root.")
else:
   print(math.sqrt(n))

9
3.0


จากตัวอย่างนี้คือทำการรับข้อมูลแล้วแปลงให้อยู่ในรูปแบบ int เก็บไว้ที่ n เพื่อที่จะนำมาหารากที่สอง แต่การจะหารากที่สองได้นั้น n ต้องไม่เป็นค่าติดลบ ดังนั้นจึงทำการตรวจสอบเงื่อนไขก่อน หาก n < 0 จะพิมพ์คำเตือน แต่ถ้าไม่ตรงตามเงื่อนไข กล่าวคือ n >= 0 จะมาทำงานในส่วนของ else ทำการหารากที่สองแล้วแสดงผลออกมา

จากตัวอย่างนี้เราเจอคำสั่งใหม่คือ import มีการใช้งานเหมือนในภาษาจาวาคือเรียกใช้ฟังก์ชันจากภายนอกซึ่งในที่นี้คือ math มีไว้สำหรับการคำนวณที่ซับซ้อนกว่า +, -, \*, / ทั่วไป

#### คำถาม
ให้เขียนโปรแกรมคิดเกรดระดับคะแนนตามสมัยเรียนมัธยม ด้วย if และ else โดยยังไม่ใช้ elif

คำตอบ

In [49]:
score = int(input())
if score >= 80:
   print('A')
else:
   if score >=70:
      print('B')
   else:
      if score >= 60:
         print('C')
      else:
         if score >= 50:
            print('D')
         else:
            print('F')

78
B


การเขียนโดยไม่ใช้ elif จะทำให้การเขียนรหัสดูไม่สวยงามเพราะต้องย่อหน้าทุกครั้งที่มีการเลือกทำ else โดยการทำงานของรหัสนี้คือ เมื่อรับ score เข้ามาแล้ว ตรวจสอบว่า score >= 80 หรือไม่ ถ้าใช่ให้แสดงผลลัพธ์ A ถ้าไม่ใช่คือ กรณีที่ score < 80 จะถูกนำมาพิจารณาต่อใน else ซึ่งเหลือค่าตั้งแต่ 79 - 0 นำมาเปรียบเทียบต่อว่า score >= 70 หรือไม่ถ้าใช่ให้แสดงผลลัพธ์ B ถ้าไม่ใช่ก็เหมือนตัดมาพิจารณาค่าตั้งแต่ 69 - 0 ไปเรื่อยๆ จนถึงกรณีที่เหลือช่วงค่า 49 - 0 ให้แสดงผลลัพธ์ F

รหัสสามารถเขียนได้อีกแบบโดยใช้ elif ซึ่งให้ความหมายและการทำงานเดียวกันกับรหัสด้านบนแต่มีความสวยงามของรหัสมากกว่า โดย elif มีความหมายคือ ถ้าไม่ตรงเงื่อนไข if ให้มาตรวจสอบเงื่อนไขที่ elif หากไม่ตรงเงื่อนไขจะตรวจสอบ elif ถัดไป ถ้าหากไม่มี elif แล้วไม่เข้ากรณีไหนเลยก็จะเข้ามาทำที่ else

In [50]:
score = int(input())
if score >= 80:
   print('A')
elif score >=70:
   print('B')
elif score >= 60:
   print('C')
elif score >= 50:
   print('D')
else:
   print('F')

78
B


การใช้ if ไม่จำเป็นต้องใช้ร่วมกับ else หรือ elif ในความหมายก็คือ ถ้าตรงตามเงื่อนไข if ก็ให้ทำขอบข่ายของ if แต่ถ้าไม่ตรงเงื่อนไขไม่ต้องทำ เช่น

In [51]:
age = int(input())
if(age > 0) :
    print("Your age is",age)
print("Bye")

30
Your age is 30
Bye


ในตัวอย่างนี้ Bye จะถูกแสดงผลออกมาไม่ว่า age จะมีค่าเท่าไรก็ตาม แต่จะแสดงผล Your age is เมื่อ age > 0 เท่านั้น

## 2.5 ฟังก์ชัน (function)
บ่อยครั้งที่มีการเขียนรหัสอยู่ในฟังก์ชัน main ยาวมาก หรือบางครั้งมีการทำงานซ้ำๆ ในส่วนต่างๆ ของรหัสซึ่งเมื่อรหัสยิ่งยาวยิ่งอ่านยาก ถ้ามีความผิดพลาด (bug) จะแก้ไขได้ยาก ดังนั้นการเรียบเรียงความคิดการเขียนรหัสให้มองการทำงานของโปรแกรมเป็นส่วนย่อยๆ จะช่วยให้สามารถตีโจทย์ แก้ไขโจทย์ได้อย่างมีระบบและระเบียบ ง่ายแก่การทดสอบและแก้ไข 

In [52]:
import math

lx = [1,2,3,4,5]
ly = [2,4,6,8,10]
lz = [1,3,5,7,9]

sum_x = 0
for item in lx :
    sum_x = sum_x + item**2
sqrt_x = math.sqrt(sum_x)
    
sum_y = 0
for item in ly :
    sum_y = sum_y + item**2
sqrt_y = math.sqrt(sum_y)
    
sum_z = 0
for item in lz :
    sum_z = sum_z + item**2
sqrt_z = math.sqrt(sum_z)
    
print(sqrt_x + sqrt_y + sqrt_z)

35.093828039952115


จากรหัสสิ่งที่ต้องการทำคือมี รับค่าข้อมูลขาเข้าเป็น list 3 ตัว แต่ละ list นำข้อมูลแต่ละตัวมายกกำลังสอง เสร็จแล้วนำมาบวกกัน ได้ผลบวกแล้วจึงนำมาหารากที่สอง เมื่อได้รากที่สองแล้วก็เอามาบวกกันของทั้ง 3 list พบว่ามีกระบวนการที่ซ้ำกันคือ "แต่ละ list นำข้อมูลแต่ละตัวมายกกำลังสอง เสร็จแล้วนำมาบวกกัน ได้ผลบวกแล้วจึงนำมาหารากที่สอง" ซึ่งสามารถลดความซ้ำซ้อนได้ด้วยการสร้างฟังก์ชัน

In [53]:
def myFunction(l) :
    mySum = 0
    for item in l :
        mySum = mySum + item**2
    mySqrt = math.sqrt(mySum)
    return mySqrt

lx = [1,2,3,4,5]
ly = [2,4,6,8,10]
lz = [1,3,5,7,9]

sqrt_x = myFunction(lx)
sqrt_y = myFunction(ly)
sqrt_z = myFunction(lz)

print(sqrt_x + sqrt_y + sqrt_z)

35.093828039952115


รหัสนี้ดูมีระเบียบและอ่านได้ง่ายกว่าอย่างเห็นได้ชัด ในส่วนของ myFunction ซึ่งเป็นฟังก์ชันที่ประกาศให้ทำงาน "แต่ละ list นำข้อมูลแต่ละตัวมายกกำลังสอง เสร็จแล้วนำมาบวกกัน ได้ผลบวกแล้วจึงนำมาหารากที่สอง" โดยแต่ละ list นั้นรับค่าผ่านพารามิเตอร์ l แล้วคืนค่าออกมาผ่าน return mySqrt 

เมื่อต้องการเขียนฟังก์ชันเนื่องจากการทำงานแบบซ้ำๆ มีหลักคิดง่ายๆ คือ ให้นึกว่าการทำงานแบบซ้ำๆ นั้นเป็นไส้ในของฟังก์ชันแต่ส่วนที่ต้องคิดเพิ่มคือ ส่วนหัวของฟังก์ชันจะต้องประกาศอย่างไร และจะคืนค่าอย่างไร 

ส่วนหัวของฟังก์ชันต้องประกาศด้วย def ชื่อฟังก์ชัน (พารามิเตอร์) : นี่คือวากยสัมพันธ์แต่สิ่งที่เป็นหัวใจคือ จะต้องส่งพารามิเตอร์อะไรเข้าสู่ฟังก์ชัน ซึ่งยากมากหากต้องการประกาศพารามิเตอร์ที่ใช้ให้ถูกต้องในครั้งเดียว จึงควรเริ่มจากการประกาศชื่อ จัดการกับไส้ในแบบเรียบง่ายก่อน 

In [54]:
def nisitFunction() :
    lx = [1,2,3,4,5]
    sum_x = 0
    for item in lx :
        sum_x = sum_x + item**2
    sqrt_x = math.sqrt(sum_x)

# call the function
nisitFunction()

ฟังก์ชันนี้เกิดจากการคัดลอกรหัสที่ยังไม่ได้จัดระเบียบมาตรงๆ จะใช้งานได้เฉพาะ lx = [1,2,3,4,5] เท่านั้น แต่ถ้าต้องการทำกับ list ใดๆ ดังนั้น list ดังกล่าวจึงควรจะเป็นหนึ่งในพารามิเตอร์ที่ส่งผ่านเข้าฟังก์ชัน

In [55]:
def nisitFunction(lx) :
    sum_x = 0
    for item in lx :
        sum_x = sum_x + item**2
    sqrt_x = math.sqrt(sum_x)
    
# call the function
lx = [1,2,3,4,5]
nisitFunction(lx)

จากรหัสด้านบนพบว่าไม่มีการคืนค่าอะไรออกมาเลย แต่จุดประสงค์ในการทำฟังก์ชันนี้ต้องการคืนค่า sqrt_x จึงเพิ่ม return sqrt_x ส่วนการเรียกใช้งานให้นำตัวแปรมารับการคืนค่าดังนี้

In [56]:
def nisitFunction(lx) :
    sum_x = 0
    for item in lx :
        sum_x = sum_x + item**2
    sqrt_x = math.sqrt(sum_x)
    return sqrt_x
    
# call the function
lx = [1,2,3,4,5]
sqrt_x = nisitFunction(lx)
print(sqrt_x)

7.416198487095663


เมื่อองค์ประกอบของฟังก์ชันครบแล้ว สิ่งที่เหลือคือนำไปใช้งานตามที่โจทย์ต้องการทั้ง lx, ly และ lz

In [57]:
import math
def nisitFunction(lx) :
    sum_x = 0
    for item in lx :
        sum_x = sum_x + item**2
    sqrt_x = math.sqrt(sum_x)
    return sqrt_x
    
# call the function
lx = [1,2,3,4,5] 
ly = [2,4,6,8,10] 
lz = [1,3,5,7,9] 

sqrt_x = nisitFunction(lx)
sqrt_y = nisitFunction(ly)
sqrt_z = nisitFunction(lz)

print(sqrt_x + sqrt_y + sqrt_z)

35.093828039952115


มีข้อสังเกตคือ nisitFunction รับ lx แต่เหตุใดเมื่อส่งค่า ly กับ lz เข้าไปถึงทำงานได้ ให้ย้อนกลับไปดูรหัสที่จัดระเบียบแล้วที่รับพารามิเตอร์เป็น l แทนที่จะรับ lx ที่ทำแบบนี้เพราะจะลดความสับสน แต่การทำงานจริงๆ แล้วให้ผลแบบเดียวกัน ณ ตรงนี้จะขออธิบาย nisitFunction คือ

บรรทัดที่ 14 sqrt_x = nisitFunction(lx) สิ่งที่เกิดขึ้นตอนเข้าไปทำ nisitFunction คือ นำ lx ที่บรรทัดที่ 10 เข้าไปกำหนดให้กับ lx ของ nisitFunction แล้วรหัสไส้ในของ nisitFunction จะเข้าใจว่า lx นี้คือ lx ที่บรรทัดที่ 10

บรรทัดที่ 15 sqrt_y = nisitFunction(ly) สิ่งที่เกิดขึ้นตอนเข้าไปทำ nisitFunction คือ นำ ly ที่บรรทัดที่ 11 เข้าไปกำหนดให้กับ lx ของ nisitFunction แล้วรหัสไส้ในของ nisitFunction จะเข้าใจว่า lx นี้คือ ly ที่บรรทัดที่ 11

บรรทัดที่ 16 sqrt_z = nisitFunction(lz) สิ่งที่เกิดขึ้นตอนเข้าไปทำ nisitFunction คือ นำ lz ที่บรรทัดที่ 12 เข้าไปกำหนดให้กับ lx ของ nisitFunction แล้วรหัสไส้ในของ nisitFunction จะเข้าใจว่า lx นี้คือ lz ที่บรรทัดที่ 12

เพื่อลดความสับสนในการใช้งานควรใช้พารามิเตอร์และตัวแปรอื่นๆ ในฟังก์ชันให้ชื่อเป็นกลางๆ จะทำให้เข้าใจได้ง่ายกว่า
และเพื่อเพิ่มเติมความเข้าใจให้พิจารณาดูความเปลี่ยนแปลงในการอ้างอิงที่ขั้นที่ 7, 24 และ 41

In [58]:
from IPython.display import IFrame
IFrame('http://www.pythontutor.com/iframe-embed.html#code=import%20math%0Adef%20nisitFunction%28lx%29%20%3A%0A%20%20%20%20sum_x%20%3D%200%0A%20%20%20%20for%20item%20in%20lx%20%3A%0A%20%20%20%20%20%20%20%20sum_x%20%3D%20sum_x%20%2B%20item**2%0A%20%20%20%20sqrt_x%20%3D%20math.sqrt%28sum_x%29%0A%20%20%20%20return%20sqrt_x%0A%20%20%20%20%0A%23%20call%20the%20function%0Alx%20%3D%20%5B1,2,3,4,5%5D%20%23%20line%209%0Aly%20%3D%20%5B2,4,6,8,10%5D%20%23%20line%2010%0Alz%20%3D%20%5B1,3,5,7,9%5D%20%23%20line%2011%0A%0Asqrt_x%20%3D%20nisitFunction%28lx%29%0Asqrt_y%20%3D%20nisitFunction%28ly%29%0Asqrt_z%20%3D%20nisitFunction%28lz%29%0A%0Aprint%28sqrt_x%20%2B%20sqrt_y%20%2B%20sqrt_z%29&cumulative=false&curInstr=6&heapPrimitives=false&mode=display&origin=opt-frontend.js&py=3&rawInputLstJSON=%5B%5D&textReferences=false', width=1050, height=512)

นอกจากนี้ฟังก์ชันยังสามารถรับพารามิเตอร์ได้มากกว่า 1 ตัว เช่น ถ้าต้องการสร้างฟังก์ชันที่บวกทุกๆ ตัวใน list l ด้วยค่า a

In [59]:
def nisitGetMoreScore(l,a) :
    for i in range(len(l)) :
        l[i] = l[i] + a
        
lx = [1,2,3,4,5]
ax = 3

nisitGetMoreScore(lx,ax)
print(lx)
    

[4, 5, 6, 7, 8]


In [60]:
from IPython.display import IFrame
IFrame('http://www.pythontutor.com/iframe-embed.html#code=def%20nisitGetMoreScore%28l,a%29%20%3A%0A%20%20%20%20for%20i%20in%20range%28len%28l%29%29%20%3A%0A%20%20%20%20%20%20%20%20l%5Bi%5D%20%3D%20l%5Bi%5D%20%2B%20a%0A%20%20%20%20%20%20%20%20%0Alx%20%3D%20%5B1,2,3,4,5%5D%0Aax%20%3D%203%0A%0AnisitGetMoreScore%28lx,ax%29%0Aprint%28lx%29%0A&cumulative=false&curInstr=4&heapPrimitives=false&mode=display&origin=opt-frontend.js&py=3&rawInputLstJSON=%5B%5D&textReferences=false', width=1050, height=512)

จากตัวอย่างนี้พบว่าฟังก์ชันนี้ไม่มีการคืนค่าและการเปลี่ยนแปลงค่าใน list สามารถทำได้ผ่านพารามิเตอร์ที่ทำได้เช่นนี้เพราะการส่งค่า lx ไปเป็น l ใน nisitGetMoreScore เป็นการส่งพารามิเตอร์แบบอ้างอิง ดังแสดงในขั้นที่ 5 ที่ lx และ l อ้างอิงไปที่ list ตัวเดียวกัน

แต่มีกรณีที่การเปลี่ยนแปลงภายในฟังก์ชันไม่ส่งผลต่อพารามิเตอร์ที่ส่งเข้าไป คือการส่งรูปแบบข้อมูลพื้นฐาน เช่น

In [61]:
def tryToAdd(b,a) :
    b = b + a
    print(b)
    
b = 1
a = 2
tryToAdd(b,a)
print(b)

3
1


In [62]:
from IPython.display import IFrame
IFrame('http://www.pythontutor.com/iframe-embed.html#code=def%20tryToAdd%28b,a%29%20%3A%0A%20%20%20%20b%20%3D%20b%20%2B%20a%0A%20%20%20%20print%28b%29%0A%20%20%20%20%0Ab%20%3D%201%0Aa%20%3D%202%0AtryToAdd%28b,a%29%0Aprint%28b%29%0A&cumulative=false&curInstr=6&heapPrimitives=false&mode=display&origin=opt-frontend.js&py=3&rawInputLstJSON=%5B%5D&textReferences=false', width=1050, height=512)

จากการไล่รหัสที่ขั้นที่ 7 จะเห็นได้ว่า b ที่เป็นตัวแปรภายใน tryToAdd ถูกเปลี่ยนแปลงค่าจาก 1 เป็น 3 ในขณะที่มีตัวแปร b ภายนอกที่ยังคงเป็น 1 อยู่เหมือนเดิม ดังนั้นในการเขียนโปรแกรมจะต้องระมัดระวังในเรื่องการส่งพารามิเตอร์แบบกำหนดค่าหรือแบบอ้างอิง เพื่อหลีกเลี่ยงความผิดพลาด

## 2.6 คำถามท้ายบท
1. ตัวดำเนินการ [:] และ [::] สำหรับ list มีการใช้งานอย่างไร
2. l[2:12:3] จะคืนค่าดัชนีใดใน l บ้าง
3. สมมติว่ามี list ซ้อน list, myList = [[1,2,3,4],[5,6,7,8],[9,10,11,12]] ถ้าต้องการจะเข้าถึงด้วยดัชนีที่ค่า 7 จะเข้าถึงได้อย่างไร
4. จงสร้าง list ที่มีค่าตั้งแต่ 5 - 15 โดยกระโดดข้ามทีละ 2 ด้วย range
5. จงสร้าง list ที่มีค่าตั้งแต่ 60 - 10 โดยลดค่าทีละ 5 ด้วย range
6. จงแสดงคำสั่งในการแปลงจาก str เป็น list และจาก list เป็น str
7. จงแสดงตัวอย่างการแปลง dict_keys ให้อยู่ในรูปของ list แล้วเรียงลำดับจากน้อยไปมาก
8. จงเขียนโปรแกรมที่ใช้โครงสร้างควบคุมในการหาค่าน้อยที่สุด ค่ามากที่สุดและค่าเฉลี่ยของ list